### DISCUSSION TITLE
Github secrets not found windows-latest / windows-2019

### DISCUSSION THREAD

**Discussion by NAJ8ry** [community/community#26988] [Actions]
<p>Followling investigation I have to conclude that currently secrets don’t work within the windows-latest or windows-2019 frameworks.</p>
<p>I have two workflows on the same repository. One uses ubuntu-latest and the other windows-latest. They should have the same visibility. When I replace the username and password on the code below with literal text it works. Secrets are not functional within the windows environment. If there is something I am missing please let me know.</p>
<p>- name: Docker login<br>
env:<br>
DockerUserName: ${{ secrets.DockerHubUser }}<br>
DockerPassword: ${{ secrets.DockerHubPassword }}<br>
run: |<br>
echo ‘Docker login’<br>
set h+</p>
<p>echo “$DockerPassword” | docker login -u “$DockerUserName” --password-stdin <a href="http://index.docker.io" rel="nofollow noopener">index.docker.io</a></p>

**Comment by NAJ8ry** [comment]
<p>Solution was to use:</p>
<p>echo “$env:DockerPassword” | docker login -u “$env:DockerUserName” --password-stdin <a href="http://index.docker.io" rel="nofollow noopener">index.docker.io</a><br>
Seems the envirnomental vars are handled slightly differently.<br>
Thank you <a class="mention" href="/ericsciple">@ericsciple</a> for your help.</p>

**Comment by Yanjingzhu** [comment]
<p>Hi <a class="mention" href="/naj8ry">@naj8ry</a> , </p>
<p>Yeah, you are right. In windows runner, the default shell for runs is powershell. In powershell script, you need to use $env:var_name syntax to invoke envirnoment variable. </p>
<p>There is the document for using a specific shell: <a href="https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#using-a-specific-shell" rel="nofollow noopener">https://help.github.com/en/actions/reference/workflow-syntax-for-github-actions#using-a-specific-shell</a> </p>

**Comment by BrightRan** [comment, marked as answer]
<p><a class="mention" href="/naj8ry">@naj8ry</a> ,</p>
<p>Yeah, when you want to access the environment variables in the command lines, the expression syntaxes are different on different shells.</p>
<p>For example:</p>
<pre><code>env:
  MSG: Hello
jobs:
  print_env_var:
    name: Print environment variable
    runs-on: windows-latest
    steps:
    - name: Print environment variable with cmd
      run: echo "String matches %MSG%"
      shell: cmd

    - name: Print environment variable with powershell
      run: echo "String matches $env:MSG"
      shell: powershell

    - name: Print environment variable with bash
      run: echo "String matches $MSG"
      shell: bash

    - name: Print environment variable with pwsh
      run: echo "String matches ${env:MSG}" # Use $env:MSG also can work
      shell: pwsh

    - name: Print environment variable with python
      run: |
        import os
        print("String matches", os.environ['MSG'])
      shell: python
</code></pre>
<p>However, if the environment variables are defined by the following ways,</p>
<ol>
<li>
<p>using the " <strong>env</strong>" key to define the custom environment variables in the workflow</p>
</li>
<li>
<p>using the Workflow command (<a href="https://help.github.com/en/actions/reference/workflow-commands-for-github-actions#setting-an-environment-variable" rel="nofollow noopener"><strong>::set-env name={name}::{value}</strong></a>) to set environment variables in the workflow</p>
</li>
</ol>
<p>the environment variables will be added to the <a href="https://help.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#env-context" rel="nofollow noopener"><strong>env context</strong></a>, then you can use the expression syntax " <strong>${{ env.&lt;env name&gt; }}</strong>" to access the environment variables. This expression syntax almost can be available to all the shells.</p>
<p>For example:</p>
<pre><code>env:
  MSG: Hello
jobs:
  print_env_var:
    name: Print environment variable
    runs-on: windows-latest
    steps:
    - name: Print environment variable with cmd
      run: echo "String matches ${{ env.MSG }}"
      shell: cmd

    - name: Print environment variable with powershell
      run: echo "String matches ${{ env.MSG }}"
      shell: powershell

    - name: Print environment variable with bash
      run: echo "String matches ${{ env.MSG }}"
      shell: bash

    - name: Print environment variable with pwsh
      run: echo "String matches ${{ env.MSG }}"
      shell: pwsh

    - name: Print environment variable with python
      run: print("String matches", "${{ env.MSG }}")
      shell: python
</code></pre>

**Comment by NAJ8ry** [comment]
<p>That is great guys,</p>
<p>Thank you for your help.</p>
<p><a class="mention" href="/brightran">@brightran</a> Is is possible to add this to the documentation on environmental variables? </p>
<p>I did read: <a href="https://help.github.com/en/actions/configuring-and-managing-workflows/using-environment-variables" rel="nofollow noopener">https://help.github.com/en/actions/configuring-and-managing-workflows/using-environment-variables</a></p>
<p>having this as an appendix might save someone else some time.</p>
<p>Thanks again,</p>
<p>Nicholas</p>
